### PR TITLE
[RFC] cmake: Improvements for building with IDEs

### PR DESCRIPTION
**Pull request by b-r-o-c-k**

### CODE REVIEW

**Comment by janlazo** [comment]
Can it detect environment variables for the Visual Studio Developer Prompt instead such as `VisualStudioVersion`? `VisualStudioEdition` is  not set for me in the developer prompt.

File: src/nvim/CMakeLists.txt [hunk 1]
```diff
@@ -430,39 +430,52 @@ if(WIN32)
...
   add_dependencies(nvim_runtime_deps nvim_dll_deps)
 
+  # A CMake script is used for copying the files to avoid the
+  # "command line is too long" error that occurs when Ninja tries running
+  # a command that exceeds the length limit (8191 characters) on Windows.
+  # See https://developercommunity.visualstudio.com/content/problem/212207/file-open-cmake-the-command-line-is-too-long.html
+  set(EXTERNAL_BLOBS_SCRIPT
+    "file(MAKE_DIRECTORY \"${PROJECT_BINARY_DIR}/windows_runtime_deps/platforms\")")
+  foreach(DEP_FILE IN ITEMS
+                      ca-bundle.crt
```
Comment (justinmk) [inline src/nvim/CMakeLists.txt]: much better!

File: CMakeLists.txt [hunk 2]
```diff
@@ -15,12 +15,41 @@ list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake")
...
   set(DEPS_PREFIX "${CMAKE_CURRENT_SOURCE_DIR}/.deps/usr" CACHE PATH "Path prefix for finding dependencies")
+  # When running from within CLion or Visual Studio,
+  # build bundled dependencies automatically.
+  if(NOT EXISTS ${DEPS_PREFIX}
+     AND (DEFINED ENV{CLION_IDE}
+          OR DEFINED ENV{VisualStudioEdition}))
+    message(STATUS "Building dependencies...")
+    set(DEPS_BUILD_DIR ${PROJECT_BINARY_DIR}/.deps)
+    file(MAKE_DIRECTORY ${DEPS_BUILD_DIR})
+    execute_process(
```
Comment (justinmk) [inline CMakeLists.txt]: Maybe this could be made into a target? Or a function or `.cmake` script. So that `build.ps1` can use it, perhaps.

**Comment by justinmk** [comment]
@b-r-o-c-k Thanks! I tried it, seems to work, although for some reason I could not select `bin/nvim.exe` as the "Startup target" in VS 2017.

Does this mean CLion works out-of-the-box now? What steps can we add to the wiki?

**Comment by b-r-o-c-k** [comment]
I recall having that problem before and I think I was able to fix it by using the "Delete Cache Folder" option in the CMake menu. I will look to see if a bug report has been made for it and submit one if there hasn't.

The build instructions for Visual Studio can be simplified by removing the steps about building the dependencies and selecting a release configuration. All configurations should work now because the dependencies will be built with the same CMake variables. Dependencies are also automatically built for CLion so selecting `Run → Build` from the menu is the only necessary step after opening the folder.

**Pull request merged**